designs/NuGet-ClientCertificates.md

```diff
@@ -0,0 +1,200 @@
+        </fromFile>
+    </clientCertificates>
+...
+</configuration>
+```
+
+## client-certificates command
+
+Gets, sets or lists client certificates to the NuGet configuration.
+
+Usage
+```
+nuget client-certificates <list|add|remove> [options]
+```
+
+if none of `list|add|remove` is specified, the command will default to `list`.
+
+### nuget client-certificates list [options]
+
+Lists all the client certificates in the configuration. This option will include all configured client certificates that match to specified options.
+
+- `-Check` one of `true|false` - Indicates that certificate existence must be checked. If the certificate is found, its fingerprint will be printed.
+
+- `-TargetSource` - Filter available client certificates for specific source.
+
+- `-SourceType` one of `file|storage` - Filter available client certificates by it's source type.
+
+- `-Name` `string` - Filter client certificates **from all sources** by string presence in it's name.
+
+- `-Path` `string` - Filter client certificates **from file source** by string presence in it's Path.
+
+- `-StoreLocation` [possible values](https://docs.microsoft.com/en-us/dotnet/api/system.security.cryptography.x509certificates.storelocation?view=netframework-4.8) - Filter client certificates **from storage source** by it's StoreLocation.
+
+- `-StoreName` [possible values](https://docs.microsoft.com/en-us/dotnet/api/system.security.cryptography.x509certificates.storename?view=netframework-4.8) - Filter client certificates **from storage source** by it's StoreName.
+
+- `-FindType` [possible values](https://docs.microsoft.com/en-us/dotnet/api/system.security.cryptography.x509certificates.x509findtype?view=netframework-4.8) - Filter client certificates **from storage source** by it's FindType.
+
+- `-FindValue` `string` - Filter client certificates from storage source by string presence in it's FindValue.
+
+Below is an example output from this command:
+
+```
+Registered client certificates:
+
+
+ 1.   file3 [fromFile]
+      Path: d:\Temp\nuget\foo.pfx
+      Password: ****
+      Certificate: 4894671AE5AA84840C31079E89E82D426BC24EC6
+
+ 2.   Foo organization certificate in storage [fromStorage]
+      Store location: CurrentUser
+      Store name: My
+      Find type: FindByThumbprint
+      Find value: ba4d3cc1f011388626a23af5627bb4721d44db6e
+      Certificate: BA4D3CC1F011388626A23AF5627BB4721D44DB6E
+```
+
+### nuget client-certificates add [options]
+
+- `-Check` one of `true|false` - Indicates that certificate existence must be checked before add action. If the certificate is not found it will not be added.
+
+- `-Name` `string` - Required option for client certificate identification. If certificate with same name exist it will be updated.
+
+- `-Path` `string` - Path to certificate file added to a file client certificate source.
+
+- `-TargetSource` `string` - Required semicolon separated option to specify NuGet sources on which client certificate will be applied.
```

Comment:
I didn't think of it during the meeting, but one problem with using semicolons on a CLI is that some shells use the semicolon to separate multiple commands written on a single line. Therefore, if we use semicolon as the separator, it increases the risk that customers will accidentally run unintended commands. While the commands would likely fail and therefore not have a large risk of causing damage, it might be better to change the design to use another separator that reduces the risk of bad commands being run accidentally.

Reply:
> Extend client-certificates command with additional required -TargetSource parameter (same multivalue behavior as -CertificateFingerprint has).

It is not mine [invention](https://docs.microsoft.com/en-us/nuget/reference/cli-reference/cli-ref-verify#examples) )

Reply:
Can a source name contain a semicolon?

Reply:
Depends on how NuGet internal NuGet.exe command line parser works. Just [reused](https://docs.microsoft.com/en-us/nuget/reference/cli-reference/cli-ref-verify#examples) existing API.